File: cpp/ml_methods/svd/demo_mlm_svd_01.cpp
// ===================================================================================
// demo_mlm_svd_01.cpp
// ===================================================================================
/**
 * @file demo_mlm_svd_01.cpp
 *
 * @brief Demonstration of SVDSolver usage.
 * Organisation:
 * -------------
 * |-- ml_methods/
 *     |-- svd/
 *         |-- demo_mlm_svd_01.cpp
 */
// ===================================================================================

// std includes
#include <iomanip>
#include <iostream>
#include <random>

// Eigen includes
#include <Eigen/Dense>

// TRex Selector includes
#include <ml_methods/svd/svd.hpp>
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>

// ============================================================================
// Namespace aliases
// ============================================================================

namespace svd_ns       = trex::ml_methods::svd;
namespace cdiagnostics = trex::utils::eval::cdiagnostics;


// ============================================================================
// Demo 1: Direct / Thin-Jacobi Path (n >= p)
// ============================================================================

void demo_svd_direct() {
    cdiagnostics::print_section_header("SVDSolver - Direct Path (n >= p)");

    // Step 1. Generate random data
    const Eigen::Index n = 300, p = 80, M = 10;
    std::cout << "Matrix size : " << n << " x " << p << "\n";
    std::cout << "M           : " << M << "\n\n";

    std::mt19937 rng(42);
    std::normal_distribution<double> norm_dist(0.0, 1.0);

    Eigen::MatrixXd X(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X(i, j) = norm_dist(rng);
        }
    }

    // Step 2. Compute SVD
    svd_ns::SVDSolver solver;
    svd_ns::SVDResult res = solver.compute(X, M);

    // Step 3. Print dimensions
    std::cout << "U dimensions : " << res.U.rows() << " x " << res.U.cols() << "\n";
    std::cout << "S size       : " << res.S.size() << "\n";
    std::cout << "V dimensions : " << res.V.rows() << " x " << res.V.cols() << "\n\n";

    // Step 4. Reconstruction error: ||X_approx - X||_F / ||X||_F
    Eigen::MatrixXd X_approx = res.U * res.S.asDiagonal() * res.V.transpose();
    double rel_error = (X_approx - X.leftCols(p)).norm() / X.norm();
    std::cout << "Relative reconstruction error ||X_M - X||_F / ||X||_F : "
              << std::scientific << rel_error << "\n";

    // Step 5. Singular values
    std::cout << "\nTop " << M << " singular values:\n";
    for (Eigen::Index k = 0; k < M; ++k) {
        std::cout << "  s" << (k + 1) << " = " << std::fixed << std::setprecision(4)
                  << res.S(k) << "\n";
    }
    std::cout << "\n\n";
}


// ============================================================================
// Demo 2: Gram / Kernel Path (p > 2*n)
// ============================================================================

void demo_svd_gram() {
    cdiagnostics::print_section_header("SVDSolver - Gram Path (p > 2*n)");

    // Step 1. Generate wide matrix
    const Eigen::Index n = 100, p = 500, M = 8;
    std::cout << "Matrix size : " << n << " x " << p << "  (p > 2*n → Gram path)\n";
    std::cout << "M           : " << M << "\n\n";

    std::mt19937 rng(7);
    std::normal_distribution<double> norm_dist(0.0, 1.0);

    Eigen::MatrixXd X(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X(i, j) = norm_dist(rng);
        }
    }

    // Step 2. Compute SVD
    svd_ns::SVDSolver solver;
    svd_ns::SVDResult res = solver.compute(X, M);

    // Step 3. Print dimensions
    std::cout << "U dimensions : " << res.U.rows() << " x " << res.U.cols() << "\n";
    std::cout << "S size       : " << res.S.size() << "\n";
    std::cout << "V dimensions : " << res.V.rows() << " x " << res.V.cols() << "\n\n";

    // Step 4. Reconstruction error
    Eigen::MatrixXd X_approx = res.U * res.S.asDiagonal() * res.V.transpose();
    double rel_error = (X_approx - X).norm() / X.norm();
    std::cout << "Relative reconstruction error ||X_M - X||_F / ||X||_F : "
              << std::scientific << rel_error << "\n\n\n";
}


// ============================================================================
// Demo 3: Orthogonality Check
// ============================================================================

void demo_svd_orthogonality() {
    cdiagnostics::print_section_header("SVDSolver - Orthogonality Check");

    const Eigen::Index n = 200, p = 150, M = 20;
    std::cout << "Matrix size : " << n << " x " << p << "\n";
    std::cout << "M           : " << M << "\n\n";

    std::mt19937 rng(99);
    std::normal_distribution<double> norm_dist(0.0, 1.0);

    Eigen::MatrixXd X(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X(i, j) = norm_dist(rng);
        }
    }

    svd_ns::SVDSolver solver;
    svd_ns::SVDResult res = solver.compute(X, M);

    // ||U^T U - I||_F
    Eigen::MatrixXd UtU = res.U.transpose() * res.U;
    double UtU_error = (UtU - Eigen::MatrixXd::Identity(M, M)).norm();

    // ||V^T V - I||_F
    Eigen::MatrixXd VtV = res.V.transpose() * res.V;
    double VtV_error = (VtV - Eigen::MatrixXd::Identity(M, M)).norm();

    std::cout << "||U^T U - I||_F : " << std::scientific << UtU_error << "\n";
    std::cout << "||V^T V - I||_F : " << std::scientific << VtV_error << "\n\n";

    if (UtU_error < 1e-10 && VtV_error < 1e-10) {
        std::cout << "✓ U and V are orthonormal\n";
    } else {
        std::cout << "✗ Orthonormality check FAILED\n";
    }

    std::cout << "\n\n";
}


// ============================================================================
// main
// ============================================================================

int main() {
    demo_svd_direct();
    demo_svd_gram();
    demo_svd_orthogonality();
    return 0;
}

File: Python/ml_methods/svd/demo_mlm_svd_01/demo_mlm_svd_01.py
# ==============================================================================
# demo_mlm_svd_01.py
# ==============================================================================
#
# Demonstration of SVDSolver usage.
#
# Mirrors cpp/ml_methods/svd/demo_mlm_svd_01/demo_mlm_svd_01.cpp.
#
# Three demos which illustrate the truncated SVD solver:
#
#   Demo 1 | Direct path   : n = 300, p = 80, M = 10 — thin-Jacobi SVD for
#                            n >= p, relative reconstruction error, top
#                            singular values (plus a full-rank M = p sanity
#                            check, a Python-side addition).
#
#   Demo 2 | Gram path     : n = 100, p = 500, M = 8 — wide matrices with
#                            p > 2*n route through the n x n Gram matrix.
#
#   Demo 3 | Orthogonality : n = 200, p = 150, M = 20 — U and V must have
#                            orthonormal columns.
#
# Notes:
#   - SVDSolver.compute(X, M) returns the rank-M truncated SVD: it keeps only
#     the M largest singular values (S), the corresponding left singular
#     vectors (U, n x M) and right singular vectors (V, p x M), so
#     U @ diag(S) @ V.T is the best rank-M approximation of X in the
#     Frobenius norm (Eckart-Young).
#   - The C++ demo seeds std::mt19937; here numpy's default_rng is seeded with
#     the same values. The RNG streams differ, so the printed numbers match
#     the C++ output qualitatively, not bitwise.
#
# ==============================================================================

import numpy as np

from trex_selector_neo.ml_methods import SVDSolver


def print_section_header(title):
    print("=" * 78)
    print(f" {title}")
    print("=" * 78 + "\n")


# ==============================================================================
# Demo 1: Direct / Thin-Jacobi Path (n >= p)
# ==============================================================================

def demo_svd_direct():
    print_section_header("SVDSolver - Direct Path (n >= p)")

    # Step 1. Generate random data
    n, p, M = 300, 80, 10
    print(f"Matrix size : {n} x {p}")
    print(f"M           : {M}\n")

    rng = np.random.default_rng(42)
    # The solver accesses the matrix through Eigen and therefore requires a
    # Fortran-ordered (column-major) float64 array.
    X = np.asfortranarray(rng.standard_normal((n, p)))

    # Step 2. Compute SVD
    solver = SVDSolver()
    res = solver.compute(X, M)

    # Step 3. Print dimensions
    print(f"U dimensions : {res.U.shape[0]} x {res.U.shape[1]}")
    print(f"S size       : {res.S.size}")
    print(f"V dimensions : {res.V.shape[0]} x {res.V.shape[1]}\n")

    # Step 4. Reconstruction error: ||X_M - X||_F / ||X||_F. With only M = 10
    #         of 80 possible components of a random Gaussian matrix, most of
    #         the energy is discarded, so this error is large by design —
    #         truncation keeps only the dominant directions.
    X_approx = res.U @ np.diag(res.S) @ res.V.T
    rel_error = np.linalg.norm(X_approx - X) / np.linalg.norm(X)
    print(f"Relative reconstruction error ||X_M - X||_F / ||X||_F : {rel_error:.6e}")

    # Step 5. Singular values (returned sorted in decreasing order)
    print(f"\nTop {M} singular values:")
    for k in range(M):
        print(f"  s{k + 1} = {res.S[k]:.4f}")

    # Python-side sanity check (not in the cpp demo): at full rank (M = p) the
    # truncated SVD reproduces X exactly up to floating-point error.
    res_full = solver.compute(X, p)
    X_full = res_full.U @ np.diag(res_full.S) @ res_full.V.T
    rel_error_full = np.linalg.norm(X_full - X) / np.linalg.norm(X)
    print(f"\nFull-rank (M = {p}) reconstruction error: {rel_error_full:.6e}")
    assert rel_error_full < 1e-10, "full-rank SVD must reconstruct X exactly"
    print("Full-rank reconstruction check passed")
    print("\n")


# ==============================================================================
# Demo 2: Gram / Kernel Path (p > 2*n)
# ==============================================================================

def demo_svd_gram():
    print_section_header("SVDSolver - Gram Path (p > 2*n)")

    # Step 1. Generate wide matrix
    n, p, M = 100, 500, 8
    print(f"Matrix size : {n} x {p}  (p > 2*n -> Gram path)")
    print(f"M           : {M}\n")

    rng = np.random.default_rng(7)
    X = np.asfortranarray(rng.standard_normal((n, p)))

    # Step 2. Compute SVD — for very wide matrices the solver works with the
    #         n x n Gram matrix X X^T instead of X itself: the same truncated
    #         factorization at far lower cost.
    solver = SVDSolver()
    res = solver.compute(X, M)

    # Step 3. Print dimensions
    print(f"U dimensions : {res.U.shape[0]} x {res.U.shape[1]}")
    print(f"S size       : {res.S.size}")
    print(f"V dimensions : {res.V.shape[0]} x {res.V.shape[1]}\n")

    # Step 4. Reconstruction error
    X_approx = res.U @ np.diag(res.S) @ res.V.T
    rel_error = np.linalg.norm(X_approx - X) / np.linalg.norm(X)
    print(f"Relative reconstruction error ||X_M - X||_F / ||X||_F : {rel_error:.6e}")
    print("\n")


# ==============================================================================
# Demo 3: Orthogonality Check
# ==============================================================================

def demo_svd_orthogonality():
    print_section_header("SVDSolver - Orthogonality Check")

    n, p, M = 200, 150, 20
    print(f"Matrix size : {n} x {p}")
    print(f"M           : {M}\n")

    rng = np.random.default_rng(99)
    X = np.asfortranarray(rng.standard_normal((n, p)))

    solver = SVDSolver()
    res = solver.compute(X, M)

    # Singular vectors are orthonormal: U^T U = I and V^T V = I (M x M).
    UtU_error = np.linalg.norm(res.U.T @ res.U - np.eye(M))
    VtV_error = np.linalg.norm(res.V.T @ res.V - np.eye(M))

    print(f"||U^T U - I||_F : {UtU_error:.6e}")
    print(f"||V^T V - I||_F : {VtV_error:.6e}\n")

    if UtU_error < 1e-10 and VtV_error < 1e-10:
        print("✓ U and V are orthonormal")
    else:
        print("✗ Orthonormality check FAILED")
    print("\n")


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    demo_svd_direct()
    demo_svd_gram()
    demo_svd_orthogonality()
